### PR TITLE
feat(aipmiddleware): implement parent validation

### PR DESCRIPTION
**Pull request by odsod**
This middleware can be used as an extra layer of defense for access
control, by ensuring that child resources are children of the parent,
for requests where a `parent` field with a `child_resource` annotation
is present.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -61,8 +62,13 @@ buf-lint: $(buf) examples/proto/api-common-protos
 	$(info [$@] linting protobuf schemas...)
 	@$(buf) lint
 
+protoc_plugins := \
+	$(protoc_gen_go) \
+	$(protoc_gen_go_grpc) \
+	$(protoc_gen_gapic_validator)
```
Comment (ericwenn) [inline Makefile:68]: What does this validate? Would it have caught invalid service?

File: aipmiddleware/validateparent.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+func (p *ParentValidator) UnaryServerInterceptor(
+	ctx context.Context,
+	request interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	protoRequest, ok := request.(proto.Message)
+	if !ok {
+		return nil, status.Error(codes.Internal, "parent validator got non-proto request")
+	}
```
Comment (ericwenn) [inline aipmiddleware/validateparent.go:31]: Not sure when this would happen, or if it is a good idea, but:
If the request is not a proto.Message, should we just pass it on instead of retuning an error?

File: aipmiddleware/validateparent.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	if parentField == nil ||
+		parentField.IsMap() ||
+		parentField.IsList() ||
+		parentField.Kind() != protoreflect.StringKind {
+		return handler(ctx, request)
+	}
+	// The parent field must have a resource reference with a child type.
+	parentResourceReference := proto.GetExtension(
+		parentField.Options(), annotations.E_ResourceReference,
+	).(*annotations.ResourceReference)
```
Comment (ericwenn) [inline aipmiddleware/validateparent.go:44]: We should handle `ok` here from the "casting", in case there is a field called parent but it does not have a resource reference.

File: aipmiddleware/validateparent.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+		return handler(ctx, request)
+	}
+	response, err := handler(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+	protoResponse, ok := response.(proto.Message)
+	if !ok {
+		return nil, status.Error(codes.Internal, "parent validator got non-proto response")
+	}
```
Comment (ericwenn) [inline aipmiddleware/validateparent.go:59]: Same here, a good idea to just pass response along instead of erroring?

File: aipmiddleware/validateparent_test.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+			tt.f(t, ctx, &tt.server, conn)
+		})
+	}
+}
+
+type mockFreightServer struct {
+	examplefreightv1.UnimplementedFreightServiceServer
+	// GetShipper
+	getShipperResponse *examplefreightv1.Shipper
+	getShipperError    error
```
Comment (ericwenn) [inline aipmiddleware/validateparent_test.go:148]: either add tests using error, or remove from struct and always return `nil`.

File: aipmiddleware/validateparent.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	protoResponse.ProtoReflect().Range(func(field protoreflect.FieldDescriptor, value protoreflect.Value) bool {
+		if field.Kind() != protoreflect.MessageKind || !field.IsList() {
```
Comment (ericwenn) [inline aipmiddleware/validateparent.go:61]: maps are also messages, but protoreflect will panic if you call `.Message` on it.

File: aipmiddleware/validateparent.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+		fieldMessageNameField := fieldMessage.Fields().ByName("name")
+		if fieldMessageNameField == nil || fieldMessageNameField.Kind() != protoreflect.StringKind {
+			return true
+		}
+		resourceDescriptor := proto.GetExtension(
+			fieldMessage.Options(), annotations.E_Resource,
+		).(*annotations.ResourceDescriptor)
+		if resourceDescriptor.GetType() != parentResourceReference.GetChildType() {
+			return true
+		}
```
Comment (ericwenn) [inline aipmiddleware/validateparent.go:74]: extract to `isResource(msg protoreflect.Message)`? This is pretty dense atm.

Comment (ericwenn) [reply aipmiddleware/validateparent.go:74]: ah, we use the child type as well. Maybe that signature needs some extra arguments. But I still think breaking some of this out is a good idea.

File: aipmiddleware/validateparent_test.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+			server: mockFreightServer{
+				getShipperResponse: &examplefreightv1.Shipper{Name: "shippers/1"},
+			},
+			f: func(t *testing.T, ctx context.Context, server *mockFreightServer, conn *grpc.ClientConn) {
+				client := examplefreightv1.NewFreightServiceClient(conn)
+				response, err := client.GetShipper(ctx, &examplefreightv1.GetShipperRequest{
+					Name: "shippers/1",
+				})
+				assert.DeepEqual(t, server.getShipperResponse, response, protocmp.Transform())
+				assert.NilError(t, err)
```
Comment (ericwenn) [inline aipmiddleware/validateparent_test.go:37]: nit: assert error first

File: aipmiddleware/validateparent_test.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+			},
+		},
+
+		{
+			name: "valid wildcard parent response",
+			server: mockFreightServer{
+				listSitesResponse: &examplefreightv1.ListSitesResponse{
+					Sites: []*examplefreightv1.Site{
+						{Name: "shippers/1/sites/1"},
+						{Name: "shippers/1/sites/2"},
```
Comment (ericwenn) [inline aipmiddleware/validateparent_test.go:68]: Make sure we properly handle multi-parent response.
```suggestion
						{Name: "shippers/2/sites/2"},
```

File: resourcename/scanner.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+	switch s.end {
+	case len(s.name):
+		return false
+	case 0:
+		// Special case for full resource names.
+		if strings.HasPrefix(s.name, "//") {
+			s.full = true
+			s.start, s.end = 2, 2
+			nextSlash := strings.IndexByte(s.name[s.start:], '/')
+			if nextSlash == -1 {
```
Comment (ericwenn) [inline resourcename/scanner.go:33]: when would this code path be triggered?

File: resourcename/hasparent_test.go [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+			parent:   "shippers/1",
+			expected: true,
+		},
+
+		{
+			test:     "empty parent",
+			name:     "shippers/1/sites/1",
+			parent:   "",
+			expected: true,
+		},
```
Comment (ericwenn) [inline resourcename/hasparent_test.go:29]: Should this be `true`? Not sure what the consequences would be, and would probably be caught by other validation in each service, but it feel like this should be `false` at first glance.

File: resourcename/hasparent_test.go [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+			parent:   "shippers/-",
+			expected: true,
+		},
+
+		{
+			test:     "full child",
+			name:     "shippers/1/sites/1",
+			parent:   "//freight-example.einride.tech/shippers/-",
+			expected: true,
+		},
```
Comment (ericwenn) [inline resourcename/hasparent_test.go:57]: test names should probably be swapped.

File: Makefile (hunk 1 above)
Comment (odsod) [reply Makefile:68]: It caught the invalid resource types

File: aipmiddleware/validateparent.go (hunk 2 above)
Comment (odsod) [reply aipmiddleware/validateparent.go:31]: gRPC uses `interface{}` since technically gRPC does not mandate protobuf. My thought here is that this middleware should only be used with proto messages, and the middleware is better off failing loudly than silently not doing anything useful.

File: aipmiddleware/validateparent.go (hunk 3 above)
Comment (odsod) [reply aipmiddleware/validateparent.go:44]: if it doesn't have a resource reference, we'll get nil here.

> GetExtension retrieves the value for an extension field. If the field is unpopulated, it returns the default value for scalars and an immutable, empty value for lists or messages. It panics if xt does not extend m.

https://pkg.go.dev/google.golang.org/protobuf/proto#GetExtension

File: aipmiddleware/validateparent.go (hunk 4 above)
Comment (odsod) [reply aipmiddleware/validateparent.go:59]: See rationale above ^

File: aipmiddleware/validateparent.go (hunk 6 above)
Comment (odsod) [reply aipmiddleware/validateparent.go:61]: Yes, but since we're requiring `field.IsList()` here we should be good, right?

File: aipmiddleware/validateparent.go (hunk 7 above)
Comment (odsod) [reply aipmiddleware/validateparent.go:74]: Agree - I'll have a look at refactoring this method

File: aipmiddleware/validateparent_test.go (hunk 8 above)
Comment (odsod) [reply aipmiddleware/validateparent_test.go:36]: Woops! Nice catch

File: aipmiddleware/validateparent_test.go (hunk 9 above)
Comment (odsod) [reply aipmiddleware/validateparent_test.go:68]: Good point! Fixing

Comment (odsod) [reply aipmiddleware/validateparent_test.go]: Fixed

File: aipmiddleware/validateparent_test.go (hunk 5 above)
Comment (odsod) [reply aipmiddleware/validateparent_test.go:178]: Good point - added more tests!

File: resourcename/hasparent_test.go (hunk 11 above)
Comment (odsod) [reply resourcename/hasparent_test.go:29]: Ah, yes let's change it to false

Comment (odsod) [reply resourcename/hasparent_test.go]: Fixed!

File: resourcename/scanner.go (hunk 10 above)
Comment (odsod) [reply resourcename/scanner.go:33]: When the resource name has a service but nothing else, so this test:

~~~go
		{
			name:        "full without segments",
			input:       "//library.googleapis.com",
			full:        true,
			serviceName: "library.googleapis.com",
			segments:    []Segment{},
		},
~~~

File: aipmiddleware/validateparent.go (hunk 7 above)
Comment (odsod) [reply aipmiddleware/validateparent.go:88]: Done!

File: Makefile (hunk 1 above)
Comment (odsod) [reply Makefile:68]: I'm thinking we can probably look at the rules in this validator, and port them over into our own linter - as part of moving the linter into here.

File: aipmiddleware/validateparent.go (hunk 2 above)
Comment (ericwenn) [reply aipmiddleware/validateparent.go]: Fair enough

File: aipmiddleware/validateparent.go (hunk 3 above)
Comment (ericwenn) [reply aipmiddleware/validateparent.go]: Amazing

File: aipmiddleware/validateparent.go (hunk 6 above)
Comment (ericwenn) [reply aipmiddleware/validateparent.go]: Well yes, but only because you can't have repeated maps. I think we should explicit with !IsMap

Comment (ericwenn) [reply aipmiddleware/validateparent.go]: Because the maps and message is so surprising in protoreflect (and proto in general) I think always checking for maps is the safest option.

Comment (odsod) [reply aipmiddleware/validateparent.go]: I see what you mean, but I think we should go by the definition of `IsList` and `IsMap` as mutually exclusive:

```go
	// IsList reports whether this field represents a list,
	// where the value type for the associated field is a List.
	// It is equivalent to checking whether Cardinality is Repeated and
	// that IsMap reports false.
```

Comment (odsod) [reply aipmiddleware/validateparent.go]: I.e. if `IsList` is true, we know by definition that `IsMap` will be false.

File: resourcename/hasparent_test.go (hunk 12 above)
Comment (odsod) [reply resourcename/hasparent_test.go]: Fixed!

File: resourcename/scanner.go (hunk 10 above)
Comment (ericwenn) [reply resourcename/scanner.go:33]: Gotcha! Had already pressed send before seeing the tests 🙉

**Review by ericwenn — APPROVED** [review]
LGTM! 
One final thought, since this potentially breaks services (for good reason), should we support logging but not returning an error as a way to roll it out? Perhaps an option "LogOnly bool".

Then we could internally roll this out in all services with LogOnly = true, give it some air time, bad then switch to errors.

**Comment by odsod** [comment]
@ericwenn I like it!!
